osu.Game.Rulesets.Osu/Edit/OsuSelectionHandler.cs

```diff
@@ -207,28 +188,93 @@ public override bool HandleRotation(float delta)
             return true;
         }
 
+        private bool scaleSlider(Slider slider, Vector2 scale)
+        {
+            Quad sliderQuad = getSurroundingQuad(slider.Path.ControlPoints.Select(p => p.Position.Value));
+            Vector2 pathRelativeDeltaScale = new Vector2(1 + scale.X / sliderQuad.Width, 1 + scale.Y / sliderQuad.Height);
+
+            Quad selectionQuad = getSurroundingQuad(new OsuHitObject[] { slider });
+            Quad scaledQuad = new Quad(selectionQuad.TopLeft.X, selectionQuad.TopLeft.Y, selectionQuad.Width + scale.X, selectionQuad.Height + scale.Y);
+            (bool X, bool Y) inBounds = isQuadInBounds(scaledQuad);
+
+            if (!inBounds.X)
+                pathRelativeDeltaScale.X = 1;
+
+            if (!inBounds.Y)
+                pathRelativeDeltaScale.Y = 1;
+
+            foreach (var point in slider.Path.ControlPoints)
+                point.Position.Value *= pathRelativeDeltaScale;
+
+            return true;
+        }
+
+        private bool scaleHitObjects(OsuHitObject[] hitObjects, Anchor reference, Vector2 scale)
+        {
+            // move the selection before scaling if dragging from top or left anchors.
+            if ((reference & Anchor.x0) > 0 && !moveSelection(new Vector2(-scale.X, 0))) return false;
+            if ((reference & Anchor.y0) > 0 && !moveSelection(new Vector2(0, -scale.Y))) return false;
```

Comment:
This doesn't make much sense now. `moveSelection` always returns true, so these guards are effectively dead - and the quad bounds are only checked after the move. This needs to be refactored at least if the guards are no longer required, but I wouldn't be surprised if there was a case that breaks this.

Reply:
Good catch. Looks like this guard has been ineffective since https://github.com/ppy/osu/commit/268bbcf77dee56a5ca1f0e4a0fb0b931c4d6f9a5, and was introduced when scaling support was added in https://github.com/ppy/osu/commit/0a10e40ce0091d40dcd0bc7e7cecebffe912ea49. 
Since `moveSelection` previously only returned `false` when trying to move hitobjects it prevented the player from scaling outside the left and top bounds(not bottom and right though).
I refactored this logic.